cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,187 @@
+                new_weights, new_args = [],[]
+                for i, (w, arg) in enumerate(zip(*lhs.args)):
+                    if is_num(arg):
+                        rhs -= w * arg
+                    else:
+                        new_weights.append(w)
+                        new_args.append(arg)
+                lhs = Operator("wsum",[new_weights, new_args])
+
+
+        if isinstance(lhs, Operator) and lhs.name == "mul" and is_num(lhs.args[0]):
+            # convert to wsum
+            lhs = Operator("wsum",[[lhs.args[0]],[lhs.args[1]]])
+
+        # now fix the comparisons themselves
+        if cpm_expr.name == "<":
+            new_rhs, cons = get_or_make_var(rhs - 1) # if rhs is constant, will return new constant
+            return [lhs <= new_rhs] + cons
+        if cpm_expr.name == ">":
+            new_rhs, cons = get_or_make_var(rhs + 1) # if rhs is constant, will return new constant
+            return [lhs >= new_rhs] + cons
+        if cpm_expr.name == "!=":
+            # Special case: BV != BV
+            if isinstance(lhs, _BoolVarImpl) and isinstance(rhs, _BoolVarImpl):
+                return [lhs + rhs == 1]
+
+            if reified or (isinstance(lhs, (Operator, GlobalConstraint)) and lhs.name not in {"sum","wsum"}):
+                # lhs is sum/wsum and rhs is contant OR
+                # lhs is GenExpr and rhs is constant or var
+                #  ... what requires less new variables?
+                # Big M implementation
+                z = boolvar()
+                # Calculate bounds of M = |lhs - rhs| + 1
+                _, M = (abs(lhs - rhs) + 1).get_bounds()
+
+                cons = [lhs - M*z <= rhs-1, lhs - M*z >= rhs-M+1]
+                return linearize_constraint(flatten_constraint(cons), supported=supported, reified=reified)
 
-    gen_constr = ["min","max","abs","pow"] #General constraints supported by gurobi, TODO: make composable for other solvers
-    if cpm_expr.name == '==':
-        lhs, rhs = cpm_expr.args
-        if lhs.is_bool() and not isinstance(lhs, _BoolVarImpl) and isinstance(rhs, _BoolVarImpl):
-            # BE == BV :: ~BV -> ~BE, BV -> BE
-            if lhs.name in gen_constr:
-                return [cpm_expr]
             else:
-                c1 = (~rhs).implies(negated_normal(lhs))
-                c2 = rhs.implies(lhs)
-                return linearize_constraint(flatten_constraint([c1,c2]))
-        elif lhs.is_bool() and not isinstance(lhs, _BoolVarImpl) and isinstance(rhs,_NumVarImpl):
-            #boolexpr == intvar, so introduce an intermediate boolvar bv, boolexpr == bv, bv == intvar
-            bv = boolvar()
-            c1 = Comparison('==', lhs, bv)
-            c2 = Comparison('==', bv, rhs)
-            return linearize_constraint(flatten_constraint([c1, c2]))
-
-    if cpm_expr.name == "xor" and len(cpm_expr.args) == 2:
-        return [sum(cpm_expr.args) == 1]
+                # introduce new indicator constraints
```

Comment:
This code seems unreachable. Missing `else:` or one less indentation?

Reply:
Also, if tests did not catch this, please add :D

Reply:
its certainly reachable and will be when the input expression is `wsum != const` for example (or != var but this var will be put to lhs by previous code)
Then we create the new constraints `aux -> wsum < const & ~aux -> wsum > const`.
Also checked using the coverage tool this case is actually triggered by the tests :)